saga_tools/slurm_run_python.py

```diff
@@ -0,0 +1,225 @@
+                stdout_log_path=job_log_directory / f"{job_name}.log",
+                spack_setup_lines="",
+                conda_base_path=self.conda_base_path,
+                conda_environment=self.conda_environment,
+                working_directory=working_directory,
+                entry_point=entry_point_name,
+                param_file=param_file.absolute(),
+            )
+            slurm_script_path.write_text(  # type: ignore
+                slurm_template_content, encoding="utf-8"
+            )
+            if echo_template:
+                print(slurm_template_content)
+            subprocess.run(
+                ["sbatch", str(slurm_script_path.absolute())],  # type: ignore
+                # Raise an exception on failure
+                check=True,
+            )
+
+    def _job_log_directory(self, job_name: str) -> Path:
+        """
+        Gets the directory to write the job logs to.
+
+        This will be `log_base_directory` unless the job name contains */*s.
+        If so, a subdirectory will be created under `log_base_directory`,
+        with each */*-separated component except the last becoming a directory level.
+
+        For example, if the job name is */foo/bar/baz*
+        and `log_base_directory` is */home/fred/logs*,
+        the returned directory will be
+        */home/fred/logs/foo/bar*.
+        """
+        parts = job_name.split("/")
+        if len(parts) > 1:
+            path_components = [self.log_base_directory]
+            # The last portion of the job name does not form part of the directory
+            # because it is used to name the log file itself.
+            path_components.extend([parts[:-1]])
+            return Path(*path_components)
+        return self.log_base_directory
+
+    _SLURM_MEMORY_UNITS = immutabledict(
+        [
+            (MemoryUnit.KILOBYTES, "K"),
+            (MemoryUnit.MEGABYTES, "M"),
+            (MemoryUnit.GIGABYTES, "G"),
+            (MemoryUnit.TERABYTES, "T"),
+        ]
+    )
+
+    def _to_slurm_memory_string(self, memory_request: MemoryAmount) -> str:
+        return (
+            f"{memory_request.amount}"
+            f"{SlurmPythonRunner._SLURM_MEMORY_UNITS[memory_request.unit]}"
+        )
+
+
+# TODO: fix hard-coded spack setup.
+# https://github.com/isi-vista/saga-tools/issues/1
+SPACK_SETUP_LINES = """
+. /scratch/spack/share/spack/setup-env.sh
+spack load cuda@9.0.176
+spack load cudnn@7.6.5.32-9.0-linux-x64
+"""
+
+SLURM_BATCH_TEMPLATE = """#!/usr/bin/env bash
+
+{account_or_qos}
+#SBATCH --partition={partition}
+#SBATCH --job-name={job_name}
+#SBATCH --output={stdout_log_path}
+#SBATCH --nodelist=gaia01
+#SBATCH --mem={memory_string}
+#SBATCH --ntasks=1
+#SBATCH --gpus-per-task={num_gpus}
+#SBATCH --cpus-per-task={num_cpus}
+
+set -euo pipefail
+
+# This is needed because SLURM jobs are run from a non-interactive shell,
+# but conda expects PS1 (the prompt variable) to be set.
+if [ -z ${{PS1+x}} ]
```

Comment:
Probably not necessary but just to avoid any potential `sh` weirdness:
```suggestion
if [[ -z ${{PS1+x}} ]]
```